skim/src/bin/main.rs

```diff
@@ -0,0 +1,237 @@
+}
+
+//------------------------------------------------------------------------------
+fn main() {
+    env_logger::builder().format_timestamp_nanos().init();
+
+    match sk_main() {
+        Ok(exit_code) => std::process::exit(exit_code),
+        Err(err) => {
+            // if downstream pipe is closed, exit silently, see PR#279
+            if err.kind() == std::io::ErrorKind::BrokenPipe {
+                std::process::exit(0)
+            }
+            std::process::exit(2)
+        }
+    }
+}
+
+fn sk_main() -> Result<i32, std::io::Error> {
+    let opts = parse_args();
+
+    let reader_opts = SkimItemReaderOption::default()
+        .ansi(opts.ansi)
+        .delimiter(&opts.delimiter)
+        .with_nth(opts.with_nth.iter().map(|s| s.as_str()))
+        .nth(opts.nth.iter().map(|s| s.as_str()))
+        .read0(opts.read0)
+        .show_error(opts.show_cmd_error);
+    let mut ctx = SkimContext {
+        cmd_collector: Rc::new(RefCell::new(SkimItemReader::new(reader_opts))),
+        query_history: vec![],
+        cmd_history: vec![],
+    };
+    ctx.init_histories(&opts);
+
+    //------------------------------------------------------------------------------
+    let bin_options = BinOptions {
+        filter: opts.filter.clone(),
+        print_query: opts.print_query,
+        print_cmd: opts.print_cmd,
+        output_ending: String::from(if opts.print0 { "\0" } else { "\n" }),
+    };
+
+    //------------------------------------------------------------------------------
+    // read from pipe or command
+    let rx_item = if !io::stdin().is_terminal() {
+        let rx_item = ctx.cmd_collector.borrow().of_bufread(BufReader::new(std::io::stdin()));
+        Some(rx_item)
+    } else {
+        None
+    };
+
+    //------------------------------------------------------------------------------
+    // filter mode
+    if opts.filter.is_some() {
+        return filter(&ctx, &bin_options, &opts, rx_item);
+    }
+
+    //------------------------------------------------------------------------------
+    // output
+
+    let Some(result) = Skim::run_with(&opts, rx_item) else {
+        return Ok(0);
+    };
+
+    if result.is_abort {
+        return Ok(130);
+    }
+
+    // output query
+    if bin_options.print_query {
+        print!("{}{}", result.query, bin_options.output_ending);
+    }
+
+    if bin_options.print_cmd {
+        print!("{}{}", result.cmd, bin_options.output_ending);
+    }
+
+    if !opts.expect.is_empty() {
+        match result.final_event {
+            Event::EvActAccept(Some(accept_key)) => {
+                print!("{}{}", accept_key, bin_options.output_ending);
+            }
+            Event::EvActAccept(None) => {
+                print!("{}", bin_options.output_ending);
+            }
+            _ => {}
+        }
+    }
+
+    for item in result.selected_items.iter() {
+        print!("{}{}", item.output(), bin_options.output_ending);
+    }
+
+    std::io::stdout().flush()?;
```

Comment:
why is this required ?

Reply:
The flush ? It is there to make sure stdout is flushed, which it might not be depending on the end delimiter.
It should not be needed, but just in case...